## WELCOME

#### This is an introductory guide to the Riot API

Have you ever wondered how sites like u.gg and op.gg get their data? \
Are you looking to become an esports analyst but not sure what skills you need? \
Want to create an app or site but not sure how to build one?

All these questions can be answered the same: The Riot API

In this short introductory guide I will take you through the process of building your first ever Riot API product \
It will require some very basic Python knowledge, but I'll explain every step along the way

I hope you enjoy, 

Jack J \
Founder & Lead Data Scientist @ iTero Gaming

### New to Python?

The tutorial doesn't require you to do any coding. Technically, you can do this all with zero knowledge \
Even so, you may want to learn more about how it works. \
There's 1000s of really good Python guides free on the internet, so I won't waste time writing another one \
I suggest going along with this tutorial and Googling certain areas where you're not sure what's happening \
If you have any questions, you could always join the Discord and ask one of the many knowledgeable persons there!

You run each cell (the grey ones with the square brackets on the left, like the one below) by first clicking into it and EITHER \
Press "Run" (top bar above) \
OR press Ctrl + Enter \
OR press Alt + Enter, which will run the cell AND create a new empty one below it

Have a test on the cell below

In [153]:
print("hello computer don't you worry i'm just learning Python")

hello computer don't you worry i'm just learning Python


### Getting a key

Head over to https://developer.riotgames.com/ \
Login with your Riot Account and accept the terms of service \
At the bottom, click "I'm not a Robot" and regenerate your API key \
Copy the API key below

In [154]:
api_key = "RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105"

### Our First API Call

Go back to the developer portal and at the top, click "APIS" \
On the left there will be a list of all the available APIs that Riot offers \
Don't be overwhelmed, it covers most of their games and we will only need a small number of them 

Scroll down until you find "Summoner-V4", once you click it you will see a list of API calls in the middle \
From here, click on the API called: /lol/summoner/v4/summoners/by-name/{summonerName} \
Scroll to the bottom of the page and type in your Summoner Name, select your Region and then select Development API Key \
Then, click "EXECUTE REQUEST" 

You should see a response code of 200, then below that the Response Body with information about your account 

Scroll back up, just below "EXECUTE REQUEST" you will see a URL that looks something like the one below \
Replace the below api_url with the one you have on the API portal


In [155]:
api_url = "https://la1.api.riotgames.com/lol/summoner/v4/summoners/by-name/M0DEST0"

We will be managing all our API requests through the "requests" library, which should be pre-installed

In [156]:
import requests

We can then send off any API request we want, so let's send off our URL

In [157]:
requests.get(api_url)

<Response [401]>

401? 

If you return to the portal and scroll up, you'll see a 401 error means: Unauthorized \
That's because we haven't attached the API key to our request!

When dealing with APIs, we tag on arguments to the request by adding "?" to the end of the url, followed by the variable

In [158]:
api_url = api_url + '?api_key=' + api_key

In [159]:
api_url

'https://la1.api.riotgames.com/lol/summoner/v4/summoners/by-name/M0DEST0?api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105'

In [160]:
requests.get(api_url)

<Response [200]>

Great! Now we're getting a 200 (which means successful) \
But where's the data? \
To get it, we need to assign the request to a variable and call the JSON

In [161]:
resp = requests.get(api_url)
player_info = resp.json()
player_info

{'id': '0U_m8GUm6deRnrPronJ5o-8WYr3d1PprBttP_qlyfBIGjg',
 'accountId': 'WCsPe0i_2CXPYDlr-GnWNdjl2ms85NKeMaweWf5lptvB2pk',
 'puuid': 'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ',
 'name': 'M0DEST0',
 'profileIconId': 5497,
 'revisionDate': 1700978571822,
 'summonerLevel': 321}

And there we go! \
We've successfully sent off and retrieved our first API request

#### Match Data

That was easy, but it's not much to show-off \
The next thing we're going to do is get some historic match data 

To do this, go back to the portal and click "MATCH-V5" on the far left \
Go the option called: /lol/match/v5/matches/by-puuid/{puuid}/ids \
As we can see from the "{puuid}" above, it needs the accounts puuid - luckily we go that from the previous API!

In [162]:
puuid = player_info['puuid']
puuid

'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ'

Scroll down on the portal until you see the box asking for the puuid \
Copy and paste the puuid above into the "puuid" box \
Leave the Query Parameters for now but update the REGION & then APP TO EXECUTE AGAINST to Development API Key \
Execute Request!

You should see a response code of 200, and the body contains a list of keys \
Let's copy the REQUEST URL from the portal so we can get everything into Python

In [163]:
api_url = "https://americas.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=20"

If you inspect the URL, you'll see there's already arguments at the end: "?start=0&count=20"\
In order to add another argument (your API key), we'll use the "&" symbol

In [164]:
#                    & instead of ?, because there's already a ? in the original
api_url = api_url + '&api_key=' + api_key 
api_url

'https://americas.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=20&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105'

Now, the same as before - let's get the data from this url

In [165]:
resp = requests.get(api_url)
match_ids = resp.json()

match_ids

['LA1_1460596933',
 'LA1_1460572182',
 'LA1_1460549970',
 'LA1_1460533557',
 'LA1_1460494945',
 'LA1_1459873188',
 'LA1_1459871292',
 'LA1_1459857257',
 'LA1_1459837219',
 'LA1_1459699435',
 'LA1_1459689802',
 'LA1_1459681842',
 'LA1_1459675534',
 'LA1_1459662989',
 'LA1_1459642993',
 'LA1_1458661840',
 'LA1_1458641722',
 'LA1_1458622715',
 'LA1_1458611574',
 'LA1_1458258515']

This API contains a list of the match IDs from our accounts last 20 games 

Let's take the first one of these

In [166]:
recent_match = match_ids[0]
recent_match

'LA1_1460596933'

If we return to the portal, scroll back up and open this API: "/lol/match/v5/matches/{matchId}"

Scroll down and copy the match ID into the "matchId" input, again selecting the right region & app

If everything goes well, you'll see our trusty 200 response code and a lot of data in RESPONSE BODY

Once again, copy and paste the REQUEST URL

In [167]:
api_url = "https://americas.api.riotgames.com/lol/match/v5/matches/LA1_1459873188"

In [168]:
# This one doesn't have any arguments (see, there's no "?" above) so we'll add it with the api key
api_url = api_url + '?api_key=' + api_key

In [169]:
resp = requests.get(api_url)
match_data = resp.json()

match_data

{'metadata': {'dataVersion': '2',
  'matchId': 'LA1_1459873188',
  'participants': ['yzHjj7_866easD7HRZAJzQF_663UgPHo7Dr0oesjZdq0PiZvpcR7nudxVb7d43YRLT9kuNNWKJ_E5Q',
   'aK3T5v83pPqbbIhH0yBAesMUK8O0UGl3NoEZWLN48tZfSkw9YqnzXFjZy8i_BJfQwVyIcWOWLJ3yZQ',
   'bulof2FC0QvUjYVeUiM4vbXtL248d5C1OquqogaU993lCjCAWo5u8E_GFljnyyP0nl5-85X1Q5nZ9g',
   'fC-3lKvs5r77swYG_qolaArnr9rba2fdVrDXTn6SuIm0FSVOgxGazPPHyOx3ceVOQyyNosykqKSXYg',
   'JYPXADV1Xhnlq9y0BA4A2vTOBotFpLJP1qQW6Mt_2uVO9NXruJ7ATkZWhS52ateRHjZqn_P_3pznXw',
   'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ',
   '3GgmyGhFZN1ViTEot9NYaeNXQyMVdESpfYcpngXQFyRJv45wO5Lpo3IXLxcwRJtKmJqgzvFCtdo5vA',
   'fDbSKld9q7WtYR8UJ_Emkuwg-ec17ZGndX-hfIjE9tbSNs-88Rwg4mYMQEdFFsmPLa6vub81HtKjvA',
   'K27ooTlTnYuVRk-y8MUScFRui7P9tYN0GPoFMVJbOvGx_u-VmPyQO8RrZYTfH2Op-0h1Mej0VhEyVw',
   '-4L801fRUAH_n6nUxYhH3K3-uTDCdZRWUxKv0IQzmj1rotZkGHP6sLcYIPTfwxGWYcfRY4JdObhowQ']},
 'info': {'gameCreation': 1700814777035,
  'gameDuration': 1839,
  

That's a lot of data!
Let's break it down

In [170]:
# the two headline sections of the data
match_data.keys()

dict_keys(['metadata', 'info'])

In [171]:
# "metadata" contains the puuid for every player in the game
match_data['metadata']

# NOTE: The order in which these puuid's appear in this the "participants" list
# is the same order they will appear for data elsewhere in this dictionary (useful for later)

{'dataVersion': '2',
 'matchId': 'LA1_1459873188',
 'participants': ['yzHjj7_866easD7HRZAJzQF_663UgPHo7Dr0oesjZdq0PiZvpcR7nudxVb7d43YRLT9kuNNWKJ_E5Q',
  'aK3T5v83pPqbbIhH0yBAesMUK8O0UGl3NoEZWLN48tZfSkw9YqnzXFjZy8i_BJfQwVyIcWOWLJ3yZQ',
  'bulof2FC0QvUjYVeUiM4vbXtL248d5C1OquqogaU993lCjCAWo5u8E_GFljnyyP0nl5-85X1Q5nZ9g',
  'fC-3lKvs5r77swYG_qolaArnr9rba2fdVrDXTn6SuIm0FSVOgxGazPPHyOx3ceVOQyyNosykqKSXYg',
  'JYPXADV1Xhnlq9y0BA4A2vTOBotFpLJP1qQW6Mt_2uVO9NXruJ7ATkZWhS52ateRHjZqn_P_3pznXw',
  'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ',
  '3GgmyGhFZN1ViTEot9NYaeNXQyMVdESpfYcpngXQFyRJv45wO5Lpo3IXLxcwRJtKmJqgzvFCtdo5vA',
  'fDbSKld9q7WtYR8UJ_Emkuwg-ec17ZGndX-hfIjE9tbSNs-88Rwg4mYMQEdFFsmPLa6vub81HtKjvA',
  'K27ooTlTnYuVRk-y8MUScFRui7P9tYN0GPoFMVJbOvGx_u-VmPyQO8RrZYTfH2Op-0h1Mej0VhEyVw',
  '-4L801fRUAH_n6nUxYhH3K3-uTDCdZRWUxKv0IQzmj1rotZkGHP6sLcYIPTfwxGWYcfRY4JdObhowQ']}

In [172]:
# "info" contains lots of data about the game, like when it was created, how long it lasted etc...
match_data['info'].keys()

dict_keys(['gameCreation', 'gameDuration', 'gameEndTimestamp', 'gameId', 'gameMode', 'gameName', 'gameStartTimestamp', 'gameType', 'gameVersion', 'mapId', 'participants', 'platformId', 'queueId', 'teams', 'tournamentCode'])

In [173]:
# For instance, this is how long the game lasted:
match_data['info']['gameDuration'] / 60 # / 60 to get it into minutes

30.65

In [174]:
# Within info, participants contains a list of length 10, each with further information about the player
len(match_data['info']['participants'])

10

In [175]:
# To save time, we'll assign a variable for first player
player_data = match_data['info']['participants'][0]
player_data

{'allInPings': 0,
 'assistMePings': 0,
 'assists': 6,
 'baitPings': 0,
 'baronKills': 0,
 'basicPings': 0,
 'bountyLevel': 4,
 'challenges': {'12AssistStreakCount': 0,
  'abilityUses': 394,
  'acesBefore15Minutes': 0,
  'alliedJungleMonsterKills': 0,
  'baronBuffGoldAdvantageOverThreshold': 1,
  'baronTakedowns': 1,
  'blastConeOppositeOpponentCount': 0,
  'bountyGold': 0,
  'buffsStolen': 0,
  'completeSupportQuestInTime': 0,
  'controlWardsPlaced': 0,
  'damagePerMinute': 1493.7438949999932,
  'damageTakenOnTeamPercentage': 0.2476551301969414,
  'dancedWithRiftHerald': 0,
  'deathsByEnemyChamps': 6,
  'dodgeSkillShotsSmallWindow': 0,
  'doubleAces': 0,
  'dragonTakedowns': 1,
  'earliestBaron': 1522.9802946999998,
  'earliestDragonTakedown': 1411.9031631,
  'earlyLaningPhaseGoldExpAdvantage': 1,
  'effectiveHealAndShielding': 0,
  'elderDragonKillsWithOpposingSoul': 0,
  'elderDragonMultikills': 0,
  'enemyChampionImmobilizations': 57,
  'enemyJungleMonsterKills': 12,
  'epicMonsterK

In [176]:
# Then, we can find out information about them in the game, like which Champion they're playing...
player_data['championName']

'Yone'

In [177]:
# Maybe their KDA?
k = player_data['kills']
d = player_data['deaths']
a = player_data['assists']
print("Kills:", k)
print("Deaths:", d)
print("Assists:", a)
print("KDA:", (k + a) / d)

Kills: 17
Deaths: 6
Assists: 6
KDA: 3.8333333333333335


In [178]:
# or their role?
player_data['teamPosition']

'TOP'

However we don't want the information about any random player in that game! \
We want to get information about the player we were searching for originally, so we need to find out where they are 

To do this, we grab the list of all the participants puuids and then we use "index" to find out where our player is

In [179]:
# A list of all the participants puuids
participants = match_data['metadata']['participants']
# Now, find where in the data our players puuid is found
player_index = participants.index(puuid)
player_index

5

In [180]:
# This should match the puuid we used to search for the match IDs, go back up and check
participants[player_index]

'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ'

Let's check it works by using this index to get the players name from the data

In [181]:
# Hopefully the name below is what you inputted into the first function!
match_data['info']['participants'][player_index]['summonerName']

'M0DEST0'

Now we know where our player is, we can grab their data

In [182]:
player_data = match_data['info']['participants'][player_index]

champ = player_data['championName']
k = player_data['kills']
d = player_data['deaths']
a = player_data['assists']
win = player_data['win']

print("Champ:", champ, "Kills:", k, "Deaths:", d, "Assists:", a, "Win:", win)

Champ: Mordekaiser Kills: 3 Deaths: 9 Assists: 5 Win: False


### Gathering Data

For whatever we're building, we may want to gather lots of data about a player \
To make this easier, we'll build some functions
Let's start by looking at the API calls we've already made

In [183]:
# The first function simply gets the puuid, given a summoner name and region
# This is exactly the same as our first example, except we're building the API URL from scratch
def get_puuid(summoner_name, region, api_key):
    api_url = (
        "https://" + 
        region +
        ".api.riotgames.com/lol/summoner/v4/summoners/by-name/" +
        summoner_name +
        "?api_key=" +
        api_key
    )
    
    print(api_url)
    
    resp = requests.get(api_url)
    player_info = resp.json()
    puuid = player_info['puuid']
    return puuid  

In [184]:
summoner_name = 'M0DEST0'
region = 'LA1'

puuid = get_puuid(summoner_name, region, api_key)
puuid

https://LA1.api.riotgames.com/lol/summoner/v4/summoners/by-name/M0DEST0?api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105


'ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ'

In [185]:
# The function to get a list of all the match IDs (2nd example above) given a players puuid and mass region
def get_match_ids(puuid, mass_region, api_key):
    api_url = (
        "https://" +
        mass_region +
        ".api.riotgames.com/lol/match/v5/matches/by-puuid/" +
        puuid + 
        "/ids?start=0&count=20" + 
        "&api_key=" + 
        api_key
    )
    
    print(api_url)
    
    resp = requests.get(api_url)
    match_ids = resp.json()
    return match_ids      

In [186]:
# NOTE: region and mass_region are different
# for instance, NA1 is the North American region
# which is part of the AMERICAS mass region
# EUW1 is Europe West region, part of the EUROPE mass region
mass_region = 'AMERICAS'

match_ids = get_match_ids(puuid, mass_region, api_key)
match_ids

https://AMERICAS.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=20&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105


['LA1_1460596933',
 'LA1_1460572182',
 'LA1_1460549970',
 'LA1_1460533557',
 'LA1_1460494945',
 'LA1_1459873188',
 'LA1_1459871292',
 'LA1_1459857257',
 'LA1_1459837219',
 'LA1_1459699435',
 'LA1_1459689802',
 'LA1_1459681842',
 'LA1_1459675534',
 'LA1_1459662989',
 'LA1_1459642993',
 'LA1_1458661840',
 'LA1_1458641722',
 'LA1_1458622715',
 'LA1_1458611574',
 'LA1_1458258515']

In [187]:
# From a given match ID and mass region, get the data about the game
def get_match_data(match_id, mass_region, api_key):
    api_url = (
        "https://" + 
        mass_region + 
        ".api.riotgames.com/lol/match/v5/matches/" +
        match_id + 
        "?api_key=" + 
        api_key
    )
    
    resp = requests.get(api_url)
    match_data = resp.json()
    return match_data     

In [188]:
match_id = match_ids[0]
match_data = get_match_data(match_id, mass_region, api_key)
match_data

{'metadata': {'dataVersion': '2',
  'matchId': 'LA1_1460596933',
  'participants': ['ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ',
   '3GgmyGhFZN1ViTEot9NYaeNXQyMVdESpfYcpngXQFyRJv45wO5Lpo3IXLxcwRJtKmJqgzvFCtdo5vA',
   'cSqaesq5mi3MGix7AdDk28PwONfU5Rlxih59EEGwA3PUrqTWVzoXw2EyNDwswKuCaaVoTYgv87B-Lw',
   'FqWjePhOHsQ_5E9dQV0vTnl0Hlu-4DQ5T9otnFcaNOSs5AdzQ-l1oH8p1OX4KkR_5Mhr1obxC650AA',
   'WH1hShydsycLS8AbDlbP3n1CyhLzUhJmw6-NvjtvmgKygbXBRhkmgGyHJkPNDCihOJtSpO3f5vx4wg',
   '_knAeY3PgCMfnvEHARY8JbbdoyfRP7g3UmHmW-GH7GvzzOIqzpTLJOomvqmj186HqPSv5Zhiqi12sA',
   'A8Gmb-TcgJsDwWUABRA3gjynDsJMBrZH-SUnbDwbXekxPpIX9eKlPxSqtXTH0k4pHG_kLS-0voiX_Q',
   'nKZ_MfhlGOOaXkPNn3fy6V6noc_hHyKYHkQN7RgVu6V0git_7NBj9M1eH6UHofAFRJE0zriK_1fwHw',
   'fqGjV8P6gkhsge0WKk_uXVxMW7l17-uzfm4aukMfnwBg23nJiHQEIXma0wrxw5wBSYaotD8M_WpJDQ',
   '-c1z5kmm2qxDlHX94c3ing647J0bQmVi5aG2w57zj2eB8nWIr_Ef8s-OiHRulY7lZdZxjjclSH5ebg']},
 'info': {'gameCreation': 1700975799006,
  'gameDuration': 2499,
  

In [189]:
# Given the match data and a players puuid, return the data about just them
def find_player_data(match_data, puuid):
    participants = match_data['metadata']['participants']
    player_index = participants.index(puuid)
    player_data = match_data['info']['participants'][player_index]
    return player_data

In [190]:
find_player_data(match_data, puuid)

{'allInPings': 0,
 'assistMePings': 0,
 'assists': 9,
 'baitPings': 0,
 'baronKills': 0,
 'basicPings': 0,
 'bountyLevel': 0,
 'challenges': {'12AssistStreakCount': 0,
  'abilityUses': 150,
  'acesBefore15Minutes': 0,
  'alliedJungleMonsterKills': 0,
  'baronTakedowns': 0,
  'blastConeOppositeOpponentCount': 0,
  'bountyGold': 0,
  'buffsStolen': 0,
  'completeSupportQuestInTime': 0,
  'controlWardTimeCoverageInRiverOrEnemyHalf': 0.6762572322932532,
  'controlWardsPlaced': 1,
  'damagePerMinute': 630.2564261771992,
  'damageTakenOnTeamPercentage': 0.20753925881638197,
  'dancedWithRiftHerald': 0,
  'deathsByEnemyChamps': 12,
  'dodgeSkillShotsSmallWindow': 0,
  'doubleAces': 0,
  'dragonTakedowns': 0,
  'earlyLaningPhaseGoldExpAdvantage': 0,
  'effectiveHealAndShielding': 0,
  'elderDragonKillsWithOpposingSoul': 0,
  'elderDragonMultikills': 0,
  'enemyChampionImmobilizations': 8,
  'enemyJungleMonsterKills': 0,
  'epicMonsterKillsNearEnemyJungler': 0,
  'epicMonsterKillsWithin30Second

We haven't done anything new above, just put everything into functions so it's easier to call them later!

Now, let's say we want to look at our last 20 games and extract the data \
To do this, we simply loop through the list of match ids and perform the functions \
For my example, each time we'll save information about the Champion, KDA and Result

In [191]:
# We initialise an empty dictionary to store data for each game
data = {
    'champion': [],
    'kills': [],
    'deaths': [],
    'assists': [],
    'win': []
}

for match_id in match_ids:
    print(match_id)
    
    # run the two functions to get the player data from the match ID
    match_data = get_match_data(match_id, mass_region, api_key)
    player_data = find_player_data(match_data, puuid)
    
    # assign the variables we're interested in
    champion = player_data['championName']
    k = player_data['kills']
    d = player_data['deaths']
    a = player_data['assists']
    win = player_data['win']
     
    # add them to our dataset
    data['champion'].append(champion)
    data['kills'].append(k)
    data['deaths'].append(d)
    data['assists'].append(a)
    data['win'].append(win)    

LA1_1460596933


LA1_1460572182
LA1_1460549970
LA1_1460533557
LA1_1460494945
LA1_1459873188
LA1_1459871292
LA1_1459857257
LA1_1459837219
LA1_1459699435
LA1_1459689802
LA1_1459681842
LA1_1459675534
LA1_1459662989
LA1_1459642993
LA1_1458661840
LA1_1458641722
LA1_1458622715
LA1_1458611574
LA1_1458258515


In [192]:
# Data on the last 20 games of the account in question
data

{'champion': ['Volibear',
  'Sett',
  'Sett',
  'Volibear',
  'Teemo',
  'Mordekaiser',
  'Mordekaiser',
  'Teemo',
  'Mordekaiser',
  'Gwen',
  'Mordekaiser',
  'Warwick',
  'Amumu',
  'Belveth',
  'Volibear',
  'Katarina',
  'Caitlyn',
  'Karthus',
  'Mordekaiser',
  'Volibear'],
 'kills': [1, 6, 6, 2, 9, 3, 0, 8, 10, 5, 4, 5, 1, 2, 3, 5, 13, 10, 7, 6],
 'deaths': [12, 10, 3, 7, 9, 9, 0, 3, 10, 7, 4, 5, 8, 6, 8, 17, 22, 21, 9, 12],
 'assists': [9,
  26,
  10,
  3,
  6,
  5,
  0,
  6,
  15,
  11,
  7,
  7,
  4,
  9,
  4,
  12,
  34,
  37,
  7,
  16],
 'win': [False,
  True,
  True,
  True,
  True,
  False,
  False,
  True,
  False,
  False,
  True,
  True,
  False,
  False,
  False,
  False,
  False,
  True,
  False,
  False]}

The next step requires the "pandas" package \
You may already have this installed, if not: https://pandas.pydata.org/getting_started.html

In [193]:
import pandas as pd # the "as" part just renames it to make it quicker to type

We'll now convert our data dictionary into a dataframe

In [194]:
df = pd.DataFrame(data)
df

champion  kills  deaths  assists    win
0      Volibear      1      12        9  False
1          Sett      6      10       26   True
2          Sett      6       3       10   True
3      Volibear      2       7        3   True
4         Teemo      9       9        6   True
5   Mordekaiser      3       9        5  False
6   Mordekaiser      0       0        0  False
7         Teemo      8       3        6   True
8   Mordekaiser     10      10       15  False
9          Gwen      5       7       11  False
10  Mordekaiser      4       4        7   True
11      Warwick      5       5        7   True
12        Amumu      1       8        4  False
13      Belveth      2       6        9  False
14     Volibear      3       8        4  False
15     Katarina      5      17       12  False
16      Caitlyn     13      22       34  False
17      Karthus     10      21       37   True
18  Mordekaiser      7       9        7  False
19     Volibear      6      12       16  False

We'll also convert the previous code into a function for later use

In [195]:
def gather_all_data(puuid, match_ids, mass_region, api_key):
    # We initialise an empty dictionary to store data for each game
    data = {
        'champion': [],
        'kills': [],
        'deaths': [],
        'assists': [],
        'win': []
    }

    for match_id in match_ids:
        print(match_id)

        # run the two functions to get the player data from the match ID
        match_data = get_match_data(match_id, mass_region, api_key)
        player_data = find_player_data(match_data, puuid)

        # assign the variables we're interested in
        champion = player_data['championName']
        k = player_data['kills']
        d = player_data['deaths']
        a = player_data['assists']
        win = player_data['win']

        # add them to our dataset
        data['champion'].append(champion)
        data['kills'].append(k)
        data['deaths'].append(d)
        data['assists'].append(a)
        data['win'].append(win)    
    
    df = pd.DataFrame(data)
    
    return df

In [196]:
df = gather_all_data(puuid, match_ids, mass_region, api_key)

LA1_1460596933


LA1_1460572182
LA1_1460549970
LA1_1460533557
LA1_1460494945
LA1_1459873188
LA1_1459871292
LA1_1459857257
LA1_1459837219
LA1_1459699435
LA1_1459689802
LA1_1459681842
LA1_1459675534
LA1_1459662989
LA1_1459642993
LA1_1458661840
LA1_1458641722
LA1_1458622715
LA1_1458611574
LA1_1458258515


In [197]:
df

champion  kills  deaths  assists    win
0      Volibear      1      12        9  False
1          Sett      6      10       26   True
2          Sett      6       3       10   True
3      Volibear      2       7        3   True
4         Teemo      9       9        6   True
5   Mordekaiser      3       9        5  False
6   Mordekaiser      0       0        0  False
7         Teemo      8       3        6   True
8   Mordekaiser     10      10       15  False
9          Gwen      5       7       11  False
10  Mordekaiser      4       4        7   True
11      Warwick      5       5        7   True
12        Amumu      1       8        4  False
13      Belveth      2       6        9  False
14     Volibear      3       8        4  False
15     Katarina      5      17       12  False
16      Caitlyn     13      22       34  False
17      Karthus     10      21       37   True
18  Mordekaiser      7       9        7  False
19     Volibear      6      12       16  False

As you can see, the "win" colun is a boolean (True or False) \
To make it easier to use, we'll convert it to an int (1 or 0)

In [198]:
df['win'] = df['win'].astype(int)

In [199]:
df

champion  kills  deaths  assists  win
0      Volibear      1      12        9    0
1          Sett      6      10       26    1
2          Sett      6       3       10    1
3      Volibear      2       7        3    1
4         Teemo      9       9        6    1
5   Mordekaiser      3       9        5    0
6   Mordekaiser      0       0        0    0
7         Teemo      8       3        6    1
8   Mordekaiser     10      10       15    0
9          Gwen      5       7       11    0
10  Mordekaiser      4       4        7    1
11      Warwick      5       5        7    1
12        Amumu      1       8        4    0
13      Belveth      2       6        9    0
14     Volibear      3       8        4    0
15     Katarina      5      17       12    0
16      Caitlyn     13      22       34    0
17      Karthus     10      21       37    1
18  Mordekaiser      7       9        7    0
19     Volibear      6      12       16    0

Now we have it as a dataframe, it's much easier to do what we want, such as:

In [200]:
# Find the averages
df.mean(numeric_only=True) # numeric_only stops it trying to average the "champion" column

kills       5.3
deaths      9.1
assists    11.4
win         0.4
dtype: float64

In [201]:
# Get the averages per champion
df.groupby('champion').mean()

kills  deaths  assists   win
champion                                 
Amumu          1.0    8.00      4.0  0.00
Belveth        2.0    6.00      9.0  0.00
Caitlyn       13.0   22.00     34.0  0.00
Gwen           5.0    7.00     11.0  0.00
Karthus       10.0   21.00     37.0  1.00
Katarina       5.0   17.00     12.0  0.00
Mordekaiser    4.8    6.40      6.8  0.20
Sett           6.0    6.50     18.0  1.00
Teemo          8.5    6.00      6.0  1.00
Volibear       3.0    9.75      8.0  0.25
Warwick        5.0    5.00      7.0  1.00

In [202]:
# or maybe order your games by amount of kills
df.sort_values('kills')

champion  kills  deaths  assists  win
6   Mordekaiser      0       0        0    0
0      Volibear      1      12        9    0
12        Amumu      1       8        4    0
3      Volibear      2       7        3    1
13      Belveth      2       6        9    0
14     Volibear      3       8        4    0
5   Mordekaiser      3       9        5    0
10  Mordekaiser      4       4        7    1
15     Katarina      5      17       12    0
11      Warwick      5       5        7    1
9          Gwen      5       7       11    0
2          Sett      6       3       10    1
1          Sett      6      10       26    1
19     Volibear      6      12       16    0
18  Mordekaiser      7       9        7    0
7         Teemo      8       3        6    1
4         Teemo      9       9        6    1
8   Mordekaiser     10      10       15    0
17      Karthus     10      21       37    1
16      Caitlyn     13      22       34    0

### Adding arguments to the API calls

20 games really isn't enough to evaluate an account, and it also includes ARAMs and Normals which you may not want \
So let's go back to one of our functions and add some arguments

In [203]:
# The original function
def get_match_ids(puuid, mass_region, api_key):
    api_url = (
        "https://" +
        mass_region +
        ".api.riotgames.com/lol/match/v5/matches/by-puuid/" +
        puuid + 
        "/ids?start=0&count=20" + 
        "&api_key=" + 
        api_key
    )
    
    print(api_url)
    
    resp = requests.get(api_url)
    match_ids = resp.json()
    return match_ids      

As you can see from the original function there is already an argument for "count=20" \
So all we need to do is replace that with a variable of our choice

In [204]:
# Updated function where you can set how many games you want
def get_match_ids(puuid, mass_region, no_games, api_key):
    api_url = (
        "https://" +
        mass_region +
        ".api.riotgames.com/lol/match/v5/matches/by-puuid/" +
        puuid + 
        "/ids?start=0" + 
        "&count=" +
        str(no_games) + 
        "&api_key=" + 
        api_key
    )
    
    print(api_url)
    
    resp = requests.get(api_url)
    match_ids = resp.json()
    return match_ids      

In [205]:
no_games = 25 # Leave this as 25 for now, otherwise you may run into issues that we cover shortly!

match_ids = get_match_ids(puuid, mass_region, no_games, api_key)

print(len(match_ids))

https://AMERICAS.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=25&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105
25


But let's say we ONLY want data from a certain queue type (I'll use Ranked Solos) \
How do we know what argument to use? 

Go back to the portal, return to MATCH-V5 (on the left) and open "/lol/match/v5/matches/by-puuid/{puuid}/ids" again \

Scroll to the bottom and fill in the form again with the same puuid, region and app \
Except this time, also add "420" to the "queue" input then EXECUTE REQUEST

Copy and paste the REQUEST URL

In [206]:
"https://americas.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=25&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105"

'https://americas.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=25&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105'

As you can see, there's a new argument: "queue=420"

If you wondering where to find all the queue types, you can go here: https://static.developer.riotgames.com/docs/lol/queues.json 

Simply find the queue type you want from the description and replace the "queue" with the queueId 

Let's add this into our function

In [207]:
# Updated function where you can set which queue to take data from
def get_match_ids(puuid, mass_region, no_games, queue_id, api_key):
    api_url = (
        "https://" +
        mass_region +
        ".api.riotgames.com/lol/match/v5/matches/by-puuid/" +
        puuid + 
        "/ids?start=0" + 
        "&count=" +
        str(no_games) + 
        "&queue=" + 
        str(queue_id) + 
        "&api_key=" + 
        api_key
    )
    
    print(api_url)
    
    resp = requests.get(api_url)
    match_ids = resp.json()
    return match_ids      

In [208]:
queue_id = 420

match_ids = get_match_ids(puuid, mass_region, no_games, queue_id, api_key)
match_ids

# Experiment with changing to different queue_ids
# For ones that don't exist, the match_ids should be an empty list

https://AMERICAS.api.riotgames.com/lol/match/v5/matches/by-puuid/ug6gjzUWzBgSY9Q2M4whHHKo_oyCatFoIuhcJHWwW6whTGrSutmObCCdtLGDqihAp5xaE7QlTAXuoQ/ids?start=0&count=25&queue=420&api_key=RGAPI-3ee6f984-bd4e-43fe-a855-51c08b131105


['LA1_1459873188',
 'LA1_1459871292',
 'LA1_1459857257',
 'LA1_1459837219',
 'LA1_1452769007',
 'LA1_1452710747',
 'LA1_1452707175',
 'LA1_1452692240',
 'LA1_1451985705',
 'LA1_1445153257',
 'LA1_1445143639',
 'LA1_1445085347',
 'LA1_1444249143',
 'LA1_1443950772',
 'LA1_1443441739',
 'LA1_1443417553',
 'LA1_1443146122',
 'LA1_1441906091',
 'LA1_1441882430',
 'LA1_1440776007',
 'LA1_1439500354',
 'LA1_1439489253',
 'LA1_1439475137',
 'LA1_1439183542',
 'LA1_1439165725']

### Rate Limits

If you return to the home page of the developer portal \
https://developer.riotgames.com/ \
Under your API key you'll see: RATE LIMITS \
This is currently set to 20 per second & 100 per 2 minutes 

So what happens if we go over it? \
Let's send off the same API request over 100 times to see:

In [209]:
# NOTE, we're sending off the same match ID 200 times, which isn't useful and only for the example
# yours may break before you reach 100, depending how fast you're going through the notebook
for i in range(200):
    print("Attempt:", i)
    match_data = get_match_data(match_id, mass_region, api_key)
    print("Match Duration:", match_data['info']['gameDuration'])

Attempt: 0
Match Duration: 2718
Attempt: 1
Match Duration: 2718
Attempt: 2
Match Duration: 2718
Attempt: 3
Match Duration: 2718
Attempt: 4
Match Duration: 2718
Attempt: 5
Match Duration: 2718
Attempt: 6
Match Duration: 2718
Attempt: 7
Match Duration: 2718
Attempt: 8
Match Duration: 2718
Attempt: 9
Match Duration: 2718
Attempt: 10
Match Duration: 2718
Attempt: 11
Match Duration: 2718
Attempt: 12
Match Duration: 2718
Attempt: 13
Match Duration: 2718
Attempt: 14
Match Duration: 2718
Attempt: 15
Match Duration: 2718
Attempt: 16
Match Duration: 2718
Attempt: 17
Match Duration: 2718
Attempt: 18
Match Duration: 2718
Attempt: 19
Match Duration: 2718
Attempt: 20
Match Duration: 2718
Attempt: 21
Match Duration: 2718
Attempt: 22
Match Duration: 2718
Attempt: 23
Match Duration: 2718
Attempt: 24
Match Duration: 2718
Attempt: 25
Match Duration: 2718
Attempt: 26
Match Duration: 2718
Attempt: 27
Match Duration: 2718
Attempt: 28
Match Duration: 2718
Attempt: 29
Match Duration: 2718
Attempt: 30
Match Du

KeyError: 'info'

The code runs fine and prints the game duration UNTIL the 101st \

So let's inspect what it returned:

In [ ]:
match_data

A status_code of 429, which means we've exceeded our rate limit

The maximum amount of recent games you can request is 100

So, from start to finish, our code currently can send up to 102 requests \
1 to get the summoner names puuid \
1 to get the puuids most 100 recent games \
up to 100 to get data on those games

So, let's update that last function to handle hitting the rate limit

In [ ]:
# Original function to get match ID
def get_match_data(match_id, mass_region, api_key):
    api_url = (
        "https://" + 
        mass_region + 
        ".api.riotgames.com/lol/match/v5/matches/" +
        match_id + 
        "?api_key=" + 
        api_key
    )
    
    resp = requests.get(api_url)
    print("Status Code:", resp.status_code) # We can read the status code straight from the "resp" object
    match_data = resp.json()
    return match_data     

First, let's do the loop again except this time printing the status_code from the "resp" object at each stage

NOTE: Depending on how quickly you're going through the notebook, you may get the error before the 101st attempt \
since you are still within the 2 minute rate limit from the previous example

In [ ]:
for i in range(120):
    print("Attempt:", i)
    get_match_data(match_id, mass_region, api_key)

We can now catch the error and prevent it from breaking our code 

To do this, we'll use the "time" package, which should be preinstalled

In [ ]:
import time

It contains a useful function called "sleep", which simply pauses your function for the given amount of seconds

In [ ]:
print("Good Night...")
time.sleep(3)
print("...Good Morning")

# if you decided to change the number to 1000 and now regret it, click Kernel->Interrupt

In [ ]:
# Updated function to sleep until it's succesful
def get_match_data(match_id, mass_region, api_key):
    api_url = (
        "https://" + 
        mass_region + 
        ".api.riotgames.com/lol/match/v5/matches/" +
        match_id + 
        "?api_key=" + 
        api_key
    )
    
    # we need to add this "while" statement so that we continuously loop until it's successful
    while True:
        resp = requests.get(api_url)
        
        # whenever we see a 429, we sleep for 10 seconds and then restart from the top of the "while" loop
        if resp.status_code == 429:
            print("Rate Limit hit, sleeping for 10 seconds")
            time.sleep(10)
            # continue means start the loop again
            continue
            
        # if resp.status_code isn't 429, then we carry on to the end of the function and return the data
        match_data = resp.json()
        return match_data     

In [ ]:
# Again, depending on how quick you are running these functions you may immediately hit the sleep
# NOTE: This function takes 2 minutes to run (since it hits the rate limit and then sleeps)
for i in range(101):
    print("Attempt:", i)
    get_match_data(match_id, mass_region, api_key)

### Wrapping it all up

With that complete, we can now wrap all our functions together into one:

In [ ]:
def master_function(summoner_name, region, mass_region, no_games, queue_id, api_key):
    puuid = get_puuid(summoner_name, region, api_key)
    match_ids = get_match_ids(puuid, mass_region, no_games, queue_id, api_key)
    df = gather_all_data(puuid, match_ids, mass_region, api_key)
    return df

In [ ]:
summoner_name = "Jack J"
region = "euw1"
mass_region = "EUROPE"
no_games = 100
queue_id = 420

In [ ]:
# Takes 2 minutes to run
df = master_function(summoner_name, region, mass_region, no_games, queue_id, api_key)

Finally, let's do something useful with the information:

In [ ]:
# print some introductory stuff
print("Hello", summoner_name, "of", region.upper()) # upper simply capitalises the region
print("Here are some interesting statistics about your last 100 solo ranked games")

# create a count column
df['count'] = 1 

# the "agg" allows us to get the average of every column but sum the count                                       # see?
champ_df = df.groupby('champion').agg({'kills': 'mean', 'deaths': 'mean', 'assists': 'mean', 'win': 'mean', 'count': 'sum'})

# we reset in the index so we can still use the "champion" column
champ_df.reset_index(inplace=True)

# we limit it to only champions where you've played 2 or more games
champ_df = champ_df[champ_df['count'] >= 2]

# create a kda column
champ_df['kda'] = (champ_df['kills'] + champ_df['assists']) / champ_df['deaths']

# sort the table by KDA, starting from the highest
champ_df = champ_df.sort_values('kda', ascending=False) # ascending determines whether it's highest to lowest or vice-versa

# assign the first row and last row to a variable so we can print information about it
best_row = champ_df.iloc[0] # .iloc[0] simply takes the first row in dataframe
worst_row = champ_df.iloc[-1] # .iloc[-1] takes the last row in a dataframe

print("Your best KDA is on", best_row['champion'], "with a KDA of", best_row['kda'], "over", best_row['count'], "game/s")
print("Your worst KDA is on", worst_row['champion'], "with a KDA of", worst_row['kda'], "over", worst_row['count'], "game/s")

# sort by count instead
champ_df = champ_df.sort_values('count', ascending=False)

# get your most played champ
row = champ_df.iloc[0]

# assign and format the win rate
win_rate = row['win']
win_rate = str(round(win_rate * 100, 1)) + "%"

print("Your highest played Champion is", row['champion'], "with", row['count'], 'game/s', 
     "and an average Win Rate of", win_rate)

# finally, sort by highest kills in a game (note, not using the champ_df groupby anymore but the raw data)
highest_kills = df.sort_values('kills', ascending=False)
row = highest_kills.iloc[0]
print("Your highest kill game was with", row['champion'], "where you had", row['kills'], "kills")


Hello M0DEST0 of LA1
Here are some interesting statistics about your last 100 solo ranked games
Your best KDA is on Mordekaiser with a KDA of 2.3333333333333335 over 6 game/s
Your worst KDA is on Gwen with a KDA of 1.352941176470588 over 3 game/s
Your highest played Champion is Mordekaiser with 6 game/s and an average Win Rate of 33.3%
Your highest kill game was with Caitlyn where you had 13 kills


### What now?

Have you finished the tutorial but now have the taste for it and want to do more? 

New to Python? If this is your first time coding, I suggest doing an introductory tutorial before anything else. \
There's tonnes of really good tutorials, guides and videos out there.

If the Python side made sense to you, here's your challenge:
- Add a new argument to one of the functions, I've set-up a space at the very bottom with everything together for you \
to play with, get hacking!
- Create a new function to get a players Champion Mastery, add a line to our final output saying who the highest mastery champion is
- Jump into the "match timeline" API and get some time specific data, like how many kills you average at 7 minutes
- From the LEAGUE-V4 API, grab 100 players from a league of your choice. Build a loop that gathers data about all 100 players recent games. Find out what the average number of kills a player has at 15 minutes for each role.
- BUILD SOMETHING NEW! Share it on our Discord. 



### What's else is available?

With that, we've concluded our introduction to the Riot API

To help you think of potential ideas, I'll list a few of the popular APIs that you can use:

- How much Mastery a player has on each Champion
- In-depth game detail for every minute of the game (i.e. how much Gold/XP each player has at 12 minutes)
- In-depth objective and kill data, like who killed who, when and where
- Ranked information, such as their current rank for each queue
- Who is currently in Challenger, Grand Master & Master (& every queue below that too!)
- And much more...!

There's also further advanced topics that I haven't covered in this introduction but you should be aware of:

- Advanced error handling, 429 Rate Limits is just one of many and each requires it's own logic
- Speeding up your code! Eventually, you may decide to build something that requires a lot of data and async/multiprocessing/threading will help
- Building and hosting the frontend of the application
- Getting Riot to approve your application for public use

I've been Jack J from iTero Gaming, I've hoped you've learnt a lot and I'm excited to see what you build!




You can join our Discord to ask questions, show off what you're building or get inspiration:
https://discord.gg/DmSUDGEW6V

You can follow me on Twitter to see what I'm using the Riot API for:
https://twitter.com/JackJGaming

I write a weekly article about statistics/AI/esports/gaming, you can subscribe here: https://medium.com/subscribe/@JackJGaming

Check out the iTero Gaming website to see what we've built:
https://itero.gg/


